sw/device/silicon_creator/lib/drivers/keymgr.c

```diff
@@ -0,0 +1,95 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/drivers/keymgr.h"
+
+#include "sw/device/silicon_creator/lib/base/abs_mmio.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "keymgr_regs.h"  // Generated.
+
+static const uint32_t base = TOP_EARLGREY_KEYMGR_BASE_ADDR;
+
+/**
+ * Checks the key manager `expected_state`.
+ *
+ * This function reads and clears the status and error code registers.
+ *
+ * @return `kErrorOk` if the key manager is at the `expected_state` and the
+ * status is idle or success.
+ */
+static rom_error_t check_expected_state(uint32_t expected_state) {
+  // Read and clear the status register by writing back the read value.
+  uint32_t op_status = abs_mmio_read32(base + KEYMGR_OP_STATUS_REG_OFFSET);
+  abs_mmio_write32(base + KEYMGR_OP_STATUS_REG_OFFSET, op_status);
+
+  // Read and clear the error register by writing back the read value.
+  uint32_t error_code = abs_mmio_read32(base + KEYMGR_ERR_CODE_REG_OFFSET);
+  abs_mmio_write32(base + KEYMGR_ERR_CODE_REG_OFFSET, error_code);
+
+  uint32_t got_state = abs_mmio_read32(base + KEYMGR_WORKING_STATE_REG_OFFSET);
+  uint32_t op_status_field =
+      bitfield_field32_read(op_status, KEYMGR_OP_STATUS_STATUS_FIELD);
+  if ((op_status_field == KEYMGR_OP_STATUS_STATUS_VALUE_IDLE ||
+       op_status_field == KEYMGR_OP_STATUS_STATUS_VALUE_DONE_SUCCESS) &&
```

Comment:
I may be being paranoid here, but we would expect `KEYMGR_OP_STATUS_STATUS_VALUE_IDLE` only after PoR, unless you were planning to call `keymgr_check_creator_state()` multiple times in the mask rom for hardening. Should we add a short comment here?

Reply:
I switched to explicit OP_STATUS checks. `IDLE` is only valid out of reset. Other states require `DONE_SUCCESS`.